clear all;
close all;
clc;

addpath('../');
addpath('../external/');

options.hgtables = '../config/hgTables_b37.txt'; % human genome annotation table
options.gcdir = '/data/suzaku/yau/cnv/gc/b37/'; % directory of local GC content files
options.outdir = '/data/suzaku/yau/oncoseq/temp/'; % output directory
options.tumourStateTable = '../config/tumourStates.txt';


options.seqtype = 'cg'; % sequencing type 'cg' or 'illumina'
options.samplename = 'test-100'; % sample name
options.infile = '/data/suzaku/yau/cg/data/aaron/summary/snps-100.txt'; % input data file


options.seqtype = 'illumina'; % sequencing type 'cg' or 'illumina'
options.samplename = 'it-met-3'; % sample name
options.infile = '/data/suzaku/yau/wgs500/pileup/snp/IT_Met_3a.txt'; % input data file

options.samplename = 'FH_BLC_745T';
options.infile = '/data/suzaku/yau/wgs500/pileup/snp/FH_BLC_745T.txt'; % input data file

options.samplename = 'cll-2'; % sample name
options.infile = '/data/suzaku/yau/wgs500/cll/snp/AS_CLL_077D.txt'; % input data file
options.infile = '/data/suzaku/yau/wgs500/cll/snp/AS_CLL_003P1.txt'; % input data file

oncoseq( 	'--read_depth_range', '[15:25]', ...
			'--chr_range', '[8 11]', ...
			'--n_train', '30000', ...
			'--maxploidy', '4.5', ...
			'--minploidy', '1.5', ...
			'--tumourheterogeneity', ...
			'--tumourstatestable', options.tumourStateTable, ...
			'--hgtable', options.hgtables, ...
			'--seqtype', 'illumina', ...
			'--samplename', options.samplename, ...
			'--infile', options.infile, ...
			'--outdir', options.outdir ...
		);
		
		
%					'--training', ...
%			'--normalcontamination', ...
%			'--maxnormalcontamination', '0.5', ...
